content/asm_3.md

```diff
@@ -11,243 +14,604 @@ global _start
 		cmp rax, 2
+        ;; Jump to the 'exit' label if not equal
 		jne exit
 		;;
-		;; Do something
+		;; Otherwise, perform another action.
 		;;
 
 incRax:
+        ;; Increment the value of the rax register
 		inc rax
+        ;; Return from the incRax subroutine
 		ret
 ```
 
-Here we can see that after application runnning, rax is equal to 1. Then we call a function incRax, which increases rax value to 1, and now rax value must be 2. After this execution continues from 8 line, where we compare rax value with 2. Also as we can read in [System V AMD64 ABI](https://refspecs.linuxbase.org/elf/x86_64-abi-0.99.pdf), the first six function arguments passed in registers. They are:
+In the example above, we can see that after the program starts, the value `1` is stored in the `rax` register. Next, we call the subroutine `incRax`, which increases the value in the `rax` register by 1. After updating the `rax` register, the subroutine ends with the `ret` instruction, and execution continues with the instructions immediately following the call to the `incRax` subroutine.
+
+In addition to preserving the return address, the stack is also used to access the function parameters and local variables. As you may recall from the previous chapter, the [System V AMD64 ABI](https://refspecs.linuxbase.org/elf/x86_64-abi-0.99.pdf) document specifies that the first six function parameters are passed in registers.
 
-* `rdi` - first argument
-* `rsi` - second argument
-* `rdx` - third argument
-* `rcx` - fourth argument
-* `r8` - fifth argument
-* `r9` - sixth
+These registers are:
 
-Next arguments will be passed in stack. So if we have function like this:
+- `rdi` - used to pass the first argument to a function.
+- `rsi` - used to pass the second argument to a function.
+- `rdx` - used to pass the third argument to a function.
+- `r10` - used to pass the fourth argument to a function.
+- `r8` - used to pass the fifth argument to a function.
+- `r9` - used to pass the sixth argument to a function.
+
+Local variables are also accessed using the stack. For example, let's take a look at the following C function that doubles its parameter:
 
 ```C
-int foo(int a1, int a2, int a3, int a4, int a5, int a6, int a7)
-{
-    return (a1 + a2 - a3 - a4 + a5 - a6) * a7;
+// The __ prefix in the __dobule function name to not mix it with the `double` data type.
+int __double(int a) {
+    int two = 2;
+
+    return a * two;
 }
 ```
 
-Then first six arguments will be passed in registers, but 7 argument will be passed in stack.
+If we compile this function and take a look at the assembly output, we will see something like this:
 
-## Stack pointer
+```assembly
+__double(int):
+        ;; Preserve the base pointer
+        push    rbp
+        ;; Set the new frame base pointer
+        mov     rbp, rsp
+        ;; Put the value of the first parameter of the function from the edi register on the stack with the location rbp - 20 bytes.
```

Comment:
The GitHub preview cuts the sentence after the word "location":

<img width="914" alt="Screenshot 2025-01-12 at 10 50 42" src="https://github.com/user-attachments/assets/33e0bd78-7ddb-4696-8b67-a3da444d27c5" />

Not sure how critical it is, though 🤔 
We can see the full sentence after scrolling a bit to the right.